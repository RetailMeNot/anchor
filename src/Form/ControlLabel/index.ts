export { ControlLabel } from './ControlLabel.component';
