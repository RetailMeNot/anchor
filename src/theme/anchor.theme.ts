// ANCHOR THEME
import { radii } from './sizes.theme';
import { MODAL_THEME_KEY, MODAL_THEME } from '../Modal';

// Copied from Adapter.StandardBreakpoints mins
// todo: dedup- they should use a single source
const breakpoints = {
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
};

interface AnchorTheme {
    colors?: any;
    breakpoints?: any;
    radii?: any;
    [key: string]: any;
}

export const anchorTheme: AnchorTheme = {
    breakpoints,
    radii,
    [MODAL_THEME_KEY]: MODAL_THEME,
};
