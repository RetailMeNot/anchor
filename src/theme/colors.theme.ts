export default {
  neutrals: {
    charcoal: {
      base: '#323232',
      light: '#595959',
      dark: '#000000',
    },
    ash: {
      base: '#A6A6A6',
      light: '#D3D3D3',
      dark: '#808080',
    },
    silver: {
      base: '#F1F1F1',
      light: '#FAFAFA',
      dark: '#E7E7E7',
    },
    lavenderWhispers: {
      base: '#F7F5FC',
      light: '#FBFAFF',
      dark: '#F1EEFA',
    },
    white: {
      base: '#FFFFFF',
    },
  },
  main: {
    grapePurchase: {
      base: '#602D6C',
      light: '#91609C',
      dark: '#4B2354',
    },
    rx: {
      base: '#F23466',
      light: '#FF6B9C',
      dark: '#CC2D52',
    },
    genie: {
      base: '#4C18F3',
      light: '#7855D9',
      dark: '#42199A',
    },
  },
  accent: {
    savvyCyan: {
      base: '#0998D6',
      light: '#66CCFF',
      dark: '#0074A6',
    },
    tealBreaker: {
      base: '#00CCCD',
      light: '#00E0D1',
      dark: '#00BDBD',
    },
  },
  secondary: {
    dealEnvy: {
      base: '#14A382',
      light: '#00CDAD',
      dark: '#268068',
    },
    flashPink: {
      base: '#DF206C',
      light: '#E665A1',
      dark: '#B81D5B',
    },
    cyberMango: {
      base: '#EB663D',
      light: '#FA9F5A',
      dark: '#C74E30',
    },
    goldMine: {
      base: '#F5A623',
      light: '#FFC761',
      dark: '#CC861D',
    },
    fireSale: {
      base: '#EA4545',
      light: '#F16667',
      dark: '#CC3B3B',
    },
  },
  alert: {
    error: {
      base: '#CC3B3B',
    },
    success: {
      base: '#268068',
    },
  },
  rx: {
    purple: {
      base: '#A678B5',
    },
  },
};
