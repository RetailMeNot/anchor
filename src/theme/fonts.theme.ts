// VARIABLES
import variables from './variables.theme';

const { baseFontSize, fontFamily } = variables;

export default {
  baseFontSize,
  fontFamily,
  paragraph: {
    base: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    informational: {
      fontSize: `${baseFontSize *  0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
  },
  navigation: {
    base: {
      fontSize: `${baseFontSize * .75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    menu: {
      fontSize: `${baseFontSize * .875}px`,
      lineHeight: `${baseFontSize}px`,
      small: {
        base: {
          fontSize: `${baseFontSize * 0.75}px`,
          fontWeight: '500',
        },
        active: {
          fontWeight: '700',
        },
      },
      large: {
        base: {
          fontWeight: '600',
        },
        active: {
          fontWeight: '700',
        },
      },
    },
  },
  heading: {
    h4: {
      fontSize: `${baseFontSize * 1.5}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h3: {
      fontSize: `${baseFontSize * 1.75}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h2: {
      fontSize: `${baseFontSize * 2}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
    h1: {
      fontSize: `${baseFontSize * 2.25}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
  },
  typography: {
    caption: {
      fontSize: `${baseFontSize * 0.75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    ui: {
      fontSize: `${baseFontSize * 0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
    body: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeading: {
      fontSize: `${baseFontSize * 1.125}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeadingLarge: {
      fontSize: `${baseFontSize * 1.25}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    header: {
      small: {
        fontSize: `${baseFontSize * 1.5}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 1.75}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 2}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 2.25}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
    },
    display: {
      small: {
        fontSize: `${baseFontSize * 2.5}px`,
        lineHeight: `${baseFontSize * 3}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 3}px`,
        lineHeight: `${baseFontSize * 3.25}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 3.5}px`,
        lineHeight: `${baseFontSize * 4}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 4}px`,
        lineHeight: `${baseFontSize * 4.5}px`,
        fontWeight: 'normal',
      },
    },
  },
};
