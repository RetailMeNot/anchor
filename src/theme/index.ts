export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
