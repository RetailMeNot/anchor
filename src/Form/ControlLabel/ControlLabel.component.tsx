// REACT
import * as React from 'react';

// VENDOR
import styled, { css } from '@xstyled/styled-components';
import { th } from '@xstyled/system';
import classnames from 'classnames';

// ANCHOR
import { cloneWithProps } from '../../utils/cloneWithProps/cloneWithProps';
import { Typography } from '../../Typography/index';
import { space as spaceStyles, SpaceProps } from '../../theme/system.theme';

const StyledLabel = styled('label')<ControlLabelProps>`
    position: relative;
    display: flex;
    ${({ labelPlacement, disabled }) =>
        css({
            color: disabled ? 'text.disabled' : 'inherit',
            flexDirection: labelPlacement === 'right' ? 'row' : 'row-reverse',
            cursor: disabled ? 'not-allowed' : 'pointer',
        })}
    ${spaceStyles}
    font-family: ${th('typography.fontFamily')};
`;

interface ControlLabelProps extends SpaceProps {
    children?: any;
    className?: string;
    id?: string;

    disabled?: boolean;
    value?: string;
    label: string;
    control: React.ReactElement<any>;
    labelPlacement?: 'left' | 'right';
    labelSpacing?: string | number;
}

export const ControlLabel = ({
    id,
    label,
    className,
    labelPlacement = 'right',
    labelSpacing = '0.5rem',
    control,
    value,
    disabled,
    ...props
}: ControlLabelProps): React.ReactElement<ControlLabelProps> => (
    <StyledLabel
        className={classnames('anchor-control-label', className)}
        htmlFor={id}
        value={value}
        label={label}
        control={control}
        disabled={disabled}
        labelPlacement={labelPlacement}
        {...props}
    >
        {cloneWithProps(control, { value, disabled })}
        {
            <Typography
                marginLeft={
                    labelPlacement === 'right' ? labelSpacing : undefined
                }
                marginRight={
                    labelPlacement === 'left' ? labelSpacing : undefined
                }
            >
                {label}
            </Typography>
        }
    </StyledLabel>
);
