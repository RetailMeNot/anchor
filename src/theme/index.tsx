// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
