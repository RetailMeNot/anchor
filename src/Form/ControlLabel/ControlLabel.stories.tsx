// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import { select, text } from '@storybook/addon-knobs';
import styled from 'styled-components';
// SUBJECT
import * as README from './README.md';
import { ControlLabel } from './ControlLabel.component';
import { Radio } from '../Radio';
import { Toggle } from '../Toggle';

const StyledStory = styled.div`
    background: #f1f1f1;
    padding: 1rem;
`;

const ToggleableToggle = () => {
    const [checked, setChecked] = React.useState<boolean>(true);
    return (
        <Toggle
            checked={checked}
            onChange={() => setChecked(!checked)}
         />
    );
}

storiesOf('Components/Form/ControlLabel', module)
    .addParameters({ readme: { sidebar: README } })
    .add('Default', () =>
        React.createElement(() => {
            const [selectedValue, setSelectedValue] = React.useState('banana');

            function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
                setSelectedValue(event.target.value);
            }

            const labelPlacement = select('labelPlacement', ['left', 'right'], 'right');
            const labelSpacing = text('labelSpacing', '') || undefined;

            return (
                <StyledStory>
                    <ControlLabel
                        value="banana"
                        label="Banana"
                        labelPlacement={labelPlacement}
                        labelSpacing={labelSpacing}
                        control={
                            <Radio
                                value="banana"
                                onChange={handleChange}
                                checked={selectedValue === 'banana'}
                            />
                        }
                    />
                    <br />
                    <ControlLabel
                        value="blueberry"
                        label="Blueberry"
                        labelPlacement={labelPlacement}
                        labelSpacing={labelSpacing}
                        control={
                            <Radio
                                value="blueberry"
                                onChange={handleChange}
                                checked={selectedValue === 'blueberry'}
                            />
                        }
                    />
                    <br />
                    <ControlLabel
                        value="orange"
                        label="Orange"
                        labelPlacement={labelPlacement}
                        labelSpacing={labelSpacing}
                        control={
                            <Radio
                                value="orange"
                                onChange={handleChange}
                                checked={selectedValue === 'orange'}
                            />
                        }
                    />
                    <br />
                    <br />
                    <br />
                    <ControlLabel
                        value="orange"
                        label="Daily Deals"
                        labelPlacement={labelPlacement}
                        labelSpacing={labelSpacing}
                        control={
                            <ToggleableToggle />
                        }
                    />
                    <br />
                    <ControlLabel
                        value="orange"
                        label="Daily Deals"
                        labelPlacement={labelPlacement}
                        labelSpacing={labelSpacing}
                        control={
                            <ToggleableToggle />
                        }
                    />
                </StyledStory>
            );
        })
    );
