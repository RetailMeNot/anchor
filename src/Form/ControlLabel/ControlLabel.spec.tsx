// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { RootTheme } from '../../theme';
// COMPONENT
import { ControlLabel } from './ControlLabel.component';

describe('Component: ControlLabel', () => {
    it('should be defined', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ControlLabel value="value" label="label" control={<div />} />
            </ThemeProvider>
        );

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
