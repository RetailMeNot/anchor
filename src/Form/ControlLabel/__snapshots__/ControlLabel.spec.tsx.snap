// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ControlLabel should be defined 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-bottom: 1rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  cursor: pointer;
}

.c1 {
  box-sizing: border-box;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  font-size: 16px;
  font-weight: normal;
  line-height: 1.5rem;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: normal;
  line-height: 1.5rem;
  text-align: inherit;
  text-transform: none;
  line-height: 1.375rem;
  margin-bottom: 1rem;
  color: inherit;
}

.c1 small {
  font-size: 87.5%;
}

<label
  className="anchor-typography anchor-control-label c0 c1"
  label="label"
  value="value"
>
  <div
    className=""
    marginRight="0.5rem"
    value="value"
  />
  label
</label>
`;
